mindmeld/models/taggers/pytorch_crf.py

```diff
@@ -0,0 +1,408 @@
+    if len(sequence[0]) == 3:
+        sparse_mats, masks, labels = zip(*sequence)
+        return diag_concat_coo_tensors(sparse_mats), torch.stack(masks), torch.stack(labels)
+    elif len(sequence[0]) == 2:
+        sparse_mats, masks = zip(*sequence)
+        return diag_concat_coo_tensors(sparse_mats), torch.stack(masks)
+
+
+class Encoder:
+    def __init__(self, feature_extractor="hash", num_feats=50000):
+
+        self.feat_extractor = DictVectorizer(dtype=np.float32) if feature_extractor == "dict" else FeatureHasher(
+            n_features=num_feats, dtype=np.float32)
+        self.label_encoder = LabelEncoder()
+        self.feat_extract_type = feature_extractor
+        self.num_classes = None
+        self.classes = None
+        self.num_feats = num_feats
+        self.ready = False
+
+    def get_tensor_data(self, feat_dicts, labels=None, fit=False):
+        if labels is None:
+            labels = []
+        if fit:
+            if self.feat_extract_type == "dict":
+                comb_dict_list = [x for seq in feat_dicts for x in seq]
+                self.feat_extractor.fit(comb_dict_list)
+                self.num_feats = len(self.feat_extractor.get_feature_names())
+            if labels:
+                self.label_encoder.fit([x for l in labels for x in l])
+                self.pad_index = len(self.label_encoder.classes_) - 1
+                self.classes = self.label_encoder.classes_
+                self.num_classes = len(self.label_encoder.classes_)
+
+            self.ready = True
+        feats = []
+        encoded_labels = []
+        seq_lens = [len(x) for x in feat_dicts]
+        max_seq_len = max(seq_lens)
+
+        for i, (x, y) in enumerate(zip_longest(feat_dicts, labels)):
+
+            padded_x = x + [{}] * (max_seq_len - seq_lens[i])
+            sparse_feat = self.feat_extractor.transform(padded_x).tocoo()
+            sparse_feat_tensor = torch.sparse_coo_tensor(
+                indices=torch.as_tensor(np.stack([sparse_feat.row, sparse_feat.col])),
+                values=torch.as_tensor(sparse_feat.data), size=sparse_feat.shape)
+            feats.append(sparse_feat_tensor)
+
+            if y:
+                transformed_label = self.label_encoder.transform(y)
+                transformed_label = np.pad(transformed_label, pad_width=(0, max_seq_len - seq_lens[i]),
+                                           constant_values=self.pad_index)
+                label_tensor = torch.as_tensor(transformed_label, dtype=torch.long)
+                encoded_labels.append(label_tensor)
+        return (feats, encoded_labels, seq_lens) if encoded_labels else (feats, seq_lens)
+
+
+# pylint: disable=too-many-instance-attributes
+class TorchCrfModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.optimizer = None
+        self.encoder = None
+        self.best_model_save_path = os.path.join(USER_CONFIG_DIR, "tmp", str(uuid.uuid4()), "best_crf_model.pt")
```

Comment:
As dicussed offline, it is better to save model within the `.generated` folder of an app to avoid any unexpected issues while loading back models from disk.

Reply:
Fixed, I think

Reply:
We'll see if something breaks